examples/jsm/nodes/ShaderNode.js

```diff
@@ -208,13 +208,25 @@ export const label = ( node, name ) => {
 
 export const temp = ( node ) => nodeObject( new VarNode( nodeObject( node ) ) );
 
-const ConvertType = function ( type, valueClass = null, valueComponents = 1 ) {
+const ConvertType = function ( type, valueComponents = 1 ) {
+
+	const map = new ArrayMap();
```

Comment:
Are you sure this approach will not generate `memory leaks`?

Reply:
Unfortunately it is impossible to use WeakMap here because each param can be primitive (and most often it will actually be). Maybe it is better to implement in ArrayMap some kind of deleting keys that are not used for a long time when its size exceeds some value?

Reply:
Something like this should work, I think...

Reply:
This should also compose vectors of uniforms and variables too, just like we have in `JoinNode`.  e.g: `vec3( 1.0, colorNode )`
I would like to recommend a different approach to `ConvertType`:

1. Convert all parameters of `primitive values` and `Three.js values (Color, Vector2, ...)` to `ConstNode`.
2. Create a `JoinNode` and add all nodes.
3. Add a `ConvertNode` at the top to ensure the desired output.

Reply:
I think that we can try to follow an approach focused in `JoinNode` and remove `ArrayMap`. https://github.com/mrdoob/three.js/pull/23635#issuecomment-1073338514

Reply:
About the cache, I think that we can create a `FloatNode` to most common values like: `-1`, `-.5`, `0`, `.5`, `1`. Other values do not tend to repeat themselves as often.